File: public/public-web/src/test/java/ch/difty/scipamato/publ/misc/ParentUrlLocaleExtractorTest.java
package ch.difty.scipamato.publ.misc;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.when;

import java.util.Locale;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import ch.difty.scipamato.publ.config.ScipamatoPublicProperties;

@SuppressWarnings("SpellCheckingInspection")
@ExtendWith(MockitoExtension.class)
class ParentUrlLocaleExtractorTest {

    private static final Locale DEFAULT_LOCALE = Locale.ENGLISH;

    private LocaleExtractor localeExtractor;

    @Mock
    private ScipamatoPublicProperties propertiesMock;

    @BeforeEach
    void setUp() {
        when(propertiesMock.getDefaultLocalization()).thenReturn("en");
        localeExtractor = new ParentUrlLocaleExtractor(propertiesMock);
    }

    @Test
    void givenNullInput_returnsDefaultLocale() {
        assertThat(localeExtractor.extractLocaleFrom(null)).isEqualTo(DEFAULT_LOCALE);
    }

    @Test
    void givenGarbledInput_returnsDefaultLocale() {
        String input = "foobar";
        assertThat(localeExtractor.extractLocaleFrom(input)).isEqualTo(DEFAULT_LOCALE);
    }

    @Test
    void givenGermanParentUrl_returnsDe() {
        String input = "https://www.foo.ch/de/projects/ludok/datenbank/";
        assertThat(localeExtractor.extractLocaleFrom(input)).isEqualTo(Locale.GERMAN);
    }

    @Test
    void givenEnglishParentUrl_returnsEn() {
        String input = "https://www.foo.ch/en/projects/ludok/datenbank/";
        assertThat(localeExtractor.extractLocaleFrom(input)).isEqualTo(Locale.ENGLISH);
    }

    @Test
    void givenFrenchParentUrl_returnsFr() {
        String input = "https://www.foo.ch/fr/projects/page-daccueil/datenbank/";
        assertThat(localeExtractor.extractLocaleFrom(input)).isEqualTo(Locale.FRENCH);
    }

    @Test
    void canHandleHttpInsteadOfHttps() {
        String input = "http://www.foo.swisstph.ch/fr/projects/page-daccueil/datenbank/";
        assertThat(localeExtractor.extractLocaleFrom(input)).isEqualTo(Locale.FRENCH);
    }

    @Test
    void canHandleCaseVariance() {
        String input = "htTps://www.foo.swisstPh.ch/fR/projects/page-Daccueil/datenbank/";
        assertThat(localeExtractor.extractLocaleFrom(input)).isEqualTo(Locale.FRENCH);
    }

    @Test
    void canHandleSingleSlash() {
        String input = "https:/www.foo.swisstph.ch/fr/projects/page-daccueil/datenbank/";
        assertThat(localeExtractor.extractLocaleFrom(input)).isEqualTo(Locale.FRENCH);
    }
}

File: public/public-web/src/test/kotlin/ch/difty/scipamato/publ/misc/ParentUrlLocaleExtractorTest.kt
package ch.difty.scipamato.publ.misc

import io.mockk.every
import io.mockk.mockk
import org.amshove.kluent.shouldBeEqualTo
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.util.Locale

@Suppress("SpellCheckingInspection")
internal class ParentUrlLocaleExtractorTest {

    private lateinit var localeExtractor: LocaleExtractor

    @BeforeEach
    fun setUp() {
        localeExtractor = ParentUrlLocaleExtractor(
            mockk {
                every { defaultLocalization } returns "en"
            }
        )
    }

    @Test
    fun givenNullInput_returnsDefaultLocale() {
        localeExtractor.extractLocaleFrom(null) shouldBeEqualTo DEFAULT_LOCALE
    }

    @Test
    fun givenGarbledInput_returnsDefaultLocale() {
        val input = "foobar"
        localeExtractor.extractLocaleFrom(input) shouldBeEqualTo DEFAULT_LOCALE
    }

    @Test
    fun givenGermanParentUrl_returnsDe() {
        val input = "https://www.foo.ch/de/projects/ludok/datenbank/"
        localeExtractor.extractLocaleFrom(input) shouldBeEqualTo Locale.GERMAN
    }

    @Test
    fun givenEnglishParentUrl_returnsEn() {
        val input = "https://www.foo.ch/en/projects/ludok/datenbank/"
        localeExtractor.extractLocaleFrom(input) shouldBeEqualTo Locale.ENGLISH
    }

    @Test
    fun givenFrenchParentUrl_returnsFr() {
        val input = "https://www.foo.ch/fr/projects/page-daccueil/datenbank/"
        localeExtractor.extractLocaleFrom(input) shouldBeEqualTo Locale.FRENCH
    }

    @Test
    fun canHandleHttpInsteadOfHttps() {
        @Suppress("HttpUrlsUsage")
        val input = "http://www.foo.swisstph.ch/fr/projects/page-daccueil/datenbank/"
        localeExtractor.extractLocaleFrom(input) shouldBeEqualTo Locale.FRENCH
    }

    @Test
    fun canHandleCaseVariance() {
        val input = "htTps://www.foo.swisstPh.ch/fR/projects/page-Daccueil/datenbank/"
        localeExtractor.extractLocaleFrom(input) shouldBeEqualTo Locale.FRENCH
    }

    @Test
    fun canHandleSingleSlash() {
        val input = "https:/www.foo.swisstph.ch/fr/projects/page-daccueil/datenbank/"
        localeExtractor.extractLocaleFrom(input) shouldBeEqualTo Locale.FRENCH
    }

    companion object {
        private val DEFAULT_LOCALE = Locale.ENGLISH
    }
}
